### ISSUE TITLE
Disk-cache broken with range requests; proposed fix

### ISSUE THREAD

**Issue by erkki** [minio/minio#9689]
## Symptoms
As of latest master / 7dbfea1353fc292d20cd399305edf00e4b91f105.

Disk cache with range requests fails to store the cached file and results in an unexpected EOF through minio-client.

## Reproduction

Server (S3/MinIO credentials exported in environment)

```
export MINIO_CACHE=on
export MINIO_CACHE_AFTER=1
export MINIO_CACHE_WATERMARK_LOW=98
export MINIO_CACHE_WATERMARK_HIGH=99
export MINIO_CACHE_DRIVES="/home/erkki/minio/cache"
go run main.go gateway s3 https://s3.dualstack.eu-central-1.amazonaws.com
```

Test files uploaded to S3

```
aws s3 cp minio-test s3://<bucket>/minio-test --cache-control 'max-age=120' --region=eu-central-1
```

minio-client, downloading ranged 1 byte from file in S3

1) Empty cache, first invocation
```
BUCKET="<bucket>" KEY="erkki/minio-test" go run minioget.go
Read 1
```

The following gets made in `cache.json`. Notice the `"ranges":{"0-0":""}` part.
```
{"version":"1.0.0","stat":{"size":4,"modTime":"2020-05-23T15:57:08.275435414Z"},"checksum":{"algorithm":"highwayhash256S","blocksize":1048576},"meta":{"etag":"ba1f2511fc30423bdbb183fe33f3dd0f"},"ranges":{"0-0":""},"hits":1}%
```

2) Second invocation
```
BUCKET="<bucket>" KEY="erkki/minio-test" go run minioget.go
ERROR: logging before flag.Parse: I0523 22:58:21.819710   12298 minioget.go:32] AT 0 GOT 0 unexpected EOF
2020/05/23 22:58:21 unexpected EOF
exit status 1
```

Contents in `cache.json`. Some metadata gets added, `ranges` is gone, `hits` is not incremented.
```
{"version":"1.0.0","stat":{"size":4,"modTime":"2020-05-23T15:58:21.818118369Z"},"checksum":{"algorithm":"highwayhash256S","blocksize":1048576},"meta":{"Cache-Control":"max-age=120","Content-Type":"binary/octet-stream","X-Amz-Server-Side-Encryption":"AES256","content-type":"binary/octet-stream","etag":"ba1f2511fc30423bdbb183fe33f3dd0f","expires":"Mon, 01 Jan 0001 00:00:00 GMT"},"hits":1}
```

3) Subsequent invocations do not fail, but nothing gets stored in the cache. All requests go through to S3. `cache.json` content stays the same as in 2).
```
BUCKET="<bucket>" KEY="erkki/minio-test" ./minioget
Read 1
```

`minioget.go`
```golang
package main

import (
	"fmt"
	"github.com/golang/glog"
	"github.com/minio/minio-go/v6"
	"io"
	"log"
	"os"
)

func main() {
	s3Client, err := minio.NewWithRegion("127.0.0.1:9000", os.Getenv("AWS_ACCESS_KEY_ID"), os.Getenv("AWS_SECRET_ACCESS_KEY"), false, "eu-central-1")
	if err != nil {
		log.Fatalln(err)
	}
	opts := minio.GetObjectOptions{}
	err = opts.SetRange(0, 0)
	if err != nil {
		log.Fatalln(err)
	}
	reader, err := s3Client.GetObject(os.Getenv("BUCKET"), os.Getenv("KEY"), opts)
	if err != nil {
		log.Fatalln(err)
	}
	defer reader.Close()
	chunk := make([]byte, 1024)
	n, err := reader.Read(chunk)
	if err != nil && err != io.EOF {
		glog.Infoln( "GOT", n, err)
		log.Fatalln(err)
	}
	fmt.Println("Read", n)
}
```

## Fix

From my limited understanding of the codebase, there's a few different codepaths that hit `saveMetadata` and only one of them supplies a non-blank `rsFileName`. In the meanwhile `rs` gets passed as non-nil in most of the same paths, apart from one. Thus, within `saveMetadata` we should check *both* `rs` and `rsFileName` and only set the filename when non-blank (aka once `Put`/`putRange` gets called).

```diff
diff --git a/cmd/disk-cache-backend.go b/cmd/disk-cache-backend.go
index 612e81a3..120c13a6 100644
--- a/cmd/disk-cache-backend.go
+++ b/cmd/disk-cache-backend.go
@@ -498,10 +498,12 @@ func (c *diskCache) saveMetadata(ctx context.Context, bucket, object string, met
        }
        // increment hits
        if rs != nil {
-               if m.Ranges == nil {
-                       m.Ranges = make(map[string]string)
+               if rsFileName != "" {
+                       if m.Ranges == nil {
+                               m.Ranges = make(map[string]string)
+                       }
+                       m.Ranges[rs.String(actualSize)] = rsFileName
                }
-               m.Ranges[rs.String(actualSize)] = rsFileName
        } else {
                // this is necessary cleanup of range files if entire object is cached.
                for _, f := range m.Ranges {
```

Happy to send a pull-request if this seems like a proper fix. Seems to work so far with local testing.

**Comment by poornas** [comment]
@erkki , that's the correct fix - please send in a PR and we can take it in. Thanks for the contribution!

**Comment by harshavardhana** [comment]
Fixed by #9698
